components/tenant-fetcher/internal/authenticator/middleware.go

```diff
@@ -0,0 +1,209 @@
+package authenticator
+
+import (
+	"context"
+	"crypto/rsa"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+	"sync"
+
+	"github.com/form3tech-oss/jwt-go"
+	"github.com/kyma-incubator/compass/components/director/pkg/authenticator"
+	"github.com/lestrrat-go/jwx/jwk"
+	"github.com/pkg/errors"
+
+	"github.com/kyma-incubator/compass/components/director/pkg/apperrors"
+	"github.com/kyma-incubator/compass/components/director/pkg/log"
+)
+
+const (
+	AuthorizationHeaderKey = "Authorization"
+)
+
+type Error struct {
+	Message string `json:"message"`
+}
+
+type Authenticator struct {
+	mux                        sync.Mutex
+	cachedJWKs                 *jwk.Set
+	jwksEndpoint               string
+	zoneId                     string
+	trustedClaimPrefixes       []string
+	subscriptionCallbacksScope string
+	allowJWTSigningNone        bool
+}
+
+func New(jwksEndpoint, zoneId, subscriptionCallbacksScope string, trustedClaimPrefixes []string, allowJWTSigningNone bool) *Authenticator {
+	return &Authenticator{
+		jwksEndpoint:               jwksEndpoint,
+		zoneId:                     zoneId,
+		trustedClaimPrefixes:       trustedClaimPrefixes,
+		subscriptionCallbacksScope: subscriptionCallbacksScope,
+		allowJWTSigningNone:        allowJWTSigningNone,
+	}
+}
+
+func (a *Authenticator) Handler() func(next http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			ctx := r.Context()
+			token, err := a.getBearerToken(r)
+			if err != nil {
+				a.writeAppError(ctx, w, err, http.StatusBadRequest)
```

Comment:
Add log before return error, as you did in the next conditions.

Reply:
Addressed it. 🚀